Automattic/jetpack
# Docker: Configurable Host Ports

Use environment variables (`PORT_WORDPRESS`, `PORT_MYSQL`, ...) to configure the host ports of the docker containers.

#### Changes proposed in this Pull Request:

This patch adds a `yarn docker:config` script that dynamically creates a `docker/docker-compose.mine` config file with host ports for the various containers configured via environment variables. This new command is added to the existing `yarn docker:compose` command (which is called by most `yarn docker:*` commands).

Example:
`PORT_WORDPRESS=1234 yarn docker:up`

The way this is implemented is a bit hacky :)

#### Use Case

The WordPress container runs on host port 80, which I'm already using on my machine. This results in the following error:

```
nukta:jetpack mdawaffe$ yarn docker:up
yarn run v1.9.4
$ yarn docker:compose up
$ yarn docker:env && yarn docker:config && docker-compose -f docker/docker-compose.yml up
$ node -e "require('fs').createWriteStream( 'docker/.env', { flags: 'a' } );"
Removing jetpack_wordpress
jetpack_mysql is up-to-date
jetpack_sftp is up-to-date
maildev is up-to-date
Recreating 7d866c9eebcc_jetpack_wordpress ... error

ERROR: for 7d866c9eebcc_jetpack_wordpress  Cannot start service wordpress: driver failed programming external connectivity on endpoint jetpack_wordpress (ee43c694d27923528b4659fdf6a19fe0fc83d6e57a586cde5b826930390dedd9): Error starting userland proxy: Bind for 0.0.0.0:80: unexpected error (Failure EADDRINUSE)

ERROR: for wordpress  Cannot start service wordpress: driver failed programming external connectivity on endpoint jetpack_wordpress (ee43c694d27923528b4659fdf6a19fe0fc83d6e57a586cde5b826930390dedd9): Error starting userland proxy: Bind for 0.0.0.0:80: unexpected error (Failure EADDRINUSE)
ERROR: Encountered errors while bringing up the project.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

#### Testing instructions:

1. `PORT_WORDPRESS=1234 yarn docker:up`
2. Verify that the site is accessible at `http://localhost:1234`.
3. `yarn docker:down` (do not specify environment variable)
4. Verify that everything stops despite not specifying `PORT_WORDPRESS`

#### Proposed changelog entry for your changes:

Use environment variables (`PORT_WORDPRESS`, `PORT_MYSQL`, ...) to configure the host ports of the docker containers.
